**Is the allignment the same?** (Code generated with the help of Chat GPT)


In [5]:
# ============================================================
# Compare MUSCLE vs MAFFT alignments
# ============================================================

!pip -q install biopython

from Bio import AlignIO

# ------------------------------------------------------------
# File locations
# ------------------------------------------------------------

muscle_path = "/content/maxcc.afa"
mafft_path = "/content/P450_aln.fasta"

# ------------------------------------------------------------
# Load alignments
# ------------------------------------------------------------

muscle = AlignIO.read(muscle_path, "fasta")
mafft = AlignIO.read(mafft_path, "fasta")

# Create dictionaries so sequence order does not matter
muscle_dict = {record.id: str(record.seq) for record in muscle}
mafft_dict = {record.id: str(record.seq) for record in mafft}

# ------------------------------------------------------------
# Basic checks
# ------------------------------------------------------------

print("MUSCLE alignment")
print("  Number of sequences:", len(muscle))
print("  Alignment length:", muscle.get_alignment_length())

print("\nMAFFT alignment")
print("  Number of sequences:", len(mafft))
print("  Alignment length:", mafft.get_alignment_length())

# ------------------------------------------------------------
# Check that the same sequences are present
# ------------------------------------------------------------

muscle_ids = set(muscle_dict)
mafft_ids = set(mafft_dict)

print("\nSEQUENCE CHECK")

if muscle_ids == mafft_ids:
    print("  Same sequence IDs: YES")
else:
    print("  Same sequence IDs: NO")

    print("\nOnly in MUSCLE:")
    print(muscle_ids - mafft_ids)

    print("\nOnly in MAFFT:")
    print(mafft_ids - muscle_ids)

# Stop here if the alignments contain different sequences
if muscle_ids != mafft_ids:
    print("\nRESULT:")
    print("The alignments cannot be identical because they contain different sequences.")

else:

    # --------------------------------------------------------
    # Compare the actual aligned sequences
    # --------------------------------------------------------

    identical = []
    different = []

    for seq_id in sorted(muscle_ids):

        if muscle_dict[seq_id] == mafft_dict[seq_id]:
            identical.append(seq_id)
        else:
            different.append(seq_id)

    print("\nALIGNMENT COMPARISON")
    print("  Identical aligned sequences:", len(identical))
    print("  Different aligned sequences:", len(different))

    # --------------------------------------------------------
    # Final determination
    # --------------------------------------------------------

    if (
        len(muscle) == len(mafft)
        and muscle.get_alignment_length() == mafft.get_alignment_length()
        and len(different) == 0
    ):

        print("\n======================================")
        print("RESULT: ALIGNMENTS ARE IDENTICAL")
        print("======================================")
        print("\nThe MUSCLE and MAFFT alignments contain:")
        print("- the same sequences")
        print("- the same residues")
        print("- the same gap positions")
        print("- the same alignment length")

    else:

        print("\n======================================")
        print("RESULT: ALIGNMENTS ARE DIFFERENT")
        print("======================================")

        if muscle.get_alignment_length() != mafft.get_alignment_length():
            print(
                "\nAlignment lengths differ:"
                f"\n  MUSCLE: {muscle.get_alignment_length()}"
                f"\n  MAFFT:  {mafft.get_alignment_length()}"
            )

        if different:
            print("\nSequences with different alignments:")

            for seq_id in different:
                print(" ", seq_id)


MUSCLE alignment
  Number of sequences: 129
  Alignment length: 4132

MAFFT alignment
  Number of sequences: 129
  Alignment length: 2871

SEQUENCE CHECK
  Same sequence IDs: YES

ALIGNMENT COMPARISON
  Identical aligned sequences: 0
  Different aligned sequences: 129

RESULT: ALIGNMENTS ARE DIFFERENT

Alignment lengths differ:
  MUSCLE: 4132
  MAFFT:  2871

Sequences with different alignments:
  gi|167003806|ref|NP_001107791.1|
  gi|167466280|ref|NP_001107860.1|
  gi|189238170|ref|XP_973400.2|
  gi|189240531|ref|XP_972348.2|
  gi|189240561|ref|XP_973646.2|
  gi|189240567|ref|XP_001815543.1|
  gi|189241028|ref|XP_970987.2|
  gi|189241724|ref|XP_001809620.1|
  gi|194473697|ref|NP_001123993.1|
  gi|194473702|ref|NP_001123994.1|
  gi|270001240|gb|EEZ97687.1|
  gi|270001269|gb|EEZ97716.1|
  gi|270001275|gb|EEZ97722.1|
  gi|270002740|gb|EEZ99187.1|
  gi|270002749|gb|EEZ99196.1|
  gi|270002796|gb|EEZ99243.1|
  gi|270002891|gb|EEZ99338.1|
  gi|270002916|gb|EEZ99363.1|
  gi|270002917|gb|EEZ993

The allignments are not the same! They both contain 129 sequences. However, the allignments lengths differ and there are no identically alligned sequences. This means MUSCLE on Bridges2 and MAFFT produced different allignments.